internal/pkg/encryption/keys/kms.go

```diff
@@ -0,0 +1,165 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package keys
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption"
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption/luks"
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption/token"
+	"github.com/siderolabs/go-retry/retry"
+	"github.com/siderolabs/kms-client/api/kms"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+
+	"github.com/siderolabs/talos/internal/pkg/endpoint"
+	"github.com/siderolabs/talos/internal/pkg/smbios"
+)
+
+// KMSToken is the userdata stored in the partition token metadata.
+type KMSToken struct {
+	SealedData []byte `json:"sealedData"`
+}
+
+// KMSKeyHandler seals token using KMS service.
+type KMSKeyHandler struct {
+	key
+	kmsEndpoint string
+}
+
+// NewKMSKeyHandler creates new KMSKeyHandler.
+func NewKMSKeyHandler(key key, kmsEndpoint string) (*KMSKeyHandler, error) {
+	return &KMSKeyHandler{
+		key:         key,
+		kmsEndpoint: kmsEndpoint,
+	}, nil
+}
+
+// NewKey implements Handler interface.
+func (h *KMSKeyHandler) NewKey() (*encryption.Key, token.Token, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```

Comment:
I think `ctx` should be passed in explicitly here, we should try to build a full context chain

Reply:
that's going to be a _very_ long change and lots of places to change...

Reply:
I don't think passing a context would be that of a nightmare, but it has to be done properly, starting one-off retry loops without cancellation doesn't sound cool to me

Reply:
Yeah, I don't mind to add it.
I think I managed to do it.